[IMP] orm: parallel iter_browse

cawo-odoo opened the pull request:
###  [IMP] orm: add optional parallelism to iter_browse.__attr__()

In some cases, e.g. if it is known that calling a certain method on the model will only trigger inserts or it is clear that updates will be disjunct, such method calls can be done in parallel.

###  [IMP] orm: add optional parallelism to iter_browse.create()

Like the same support added to `__attr__` in the parent commit, this can only be used by callers when it is known that database modifications will be distinct, not causing concurrency issues or side-effects on the results.

cawo-odoo committed:
[IMP] orm: add optional parallelism to iter_browse.__attr__()

In some cases, e.g. if it is known that calling a certain method on the model
will only trigger inserts or it is clear that updates will be disjunct, such
method calls can be done in parallel.

aj-fuentes reviewed (commented):
To simplify the code I think we should check UPG_PARALLEL_ITER_BROWSE only _once_.

aj-fuentes reviewed (commented):
Proposal: get rid of the hack injecting `params`.

Code is simpler, easier to read, and less hacky.

cawo-odoo commented:
> Proposal: get rid of the hack injecting `params`.
> 
> Code is simpler, easier to read, and less hacky.

It is true that it looks that way, but there's on big disadvantage: it reduces the possible use-cases, because now everything in there (including `{kw,}args`) needs to be :cucumber:able.

aj-fuentes commented:
> > Proposal: get rid of the hack injecting `params`.
> > Code is simpler, easier to read, and less hacky.
> 
> It is true that it looks that way, but there's one big disadvantage: it reduces the possible use-cases, because now everything in there (including `{kw,}args`) needs to be 🥒able.

Do you have a use case where we need to pass more complex objects? I think once we have it we can make this hacky.

cawo-odoo commented:
> you're tying the code to the _internal implementation_ of the standard Python library.

What do you mean? The fact that it forks? This whole thing is tied to that, as are all of our other uses of ProcessPoolExecutor, that's why https://github.com/odoo/upgrade-util/pull/320#discussion_r2348147163 is important.

I also wanted simplicity: in the callback signature, to avoid as many serialisation issues as we can. OTOH, your suggestion is more readable and easier to understand. I think both are valid goals and they are only in conflict here, because `ProcessPoolExecutor.map` (not `map`) is *not* a simple interface, it just pretends to be. I made this point on the last PR already and it didn't change, there are all kinds of quirks in this thing.

Anyway. It is true that input to this code - if used - has to be chosen very carefully for all kinds of reasons. So maybe it is a good idea to restrict, in this way, the arguments that can be passed to the model attr being called, for the purpose of limiting the use-cases to simpler ones that hopefully have a lower risk of dangerous use.

aj-fuentes commented:
> What do you mean? The fact that it forks?

No, the fact that `map` would use pickle to send the parameters. 

Once more, I'm OK with this. You could probably include some comments about why you chose to do this. I'd say something like: "we want to avoid the overhead of serializing the same parameters for all ids via pickle in the map implementation. Warning: we need to clean them up in _end to avoid any potential conflicts".

Note that we need to keep in mind --some testing would be nice-- what happens for multiple calls to iter_browse with multiprocessing within the same run (upgrade step). I don't see anything problematic at the moment.

> it just pretends to be

Yes. We should honor that --if we can--, that's the point.

cawo-odoo committed:
[IMP] orm: add optional parallelism to iter_browse.create()

Like the same support added to `__attr__` in the parent commit, this can only
be used by callers when it is known that database modifications will be
distinct, not causing concurrency issues or side-effects on the results.

aj-fuentes commented on src/util/orm.py:
```diff
@@ -476,31 +616,72 @@ def create(self, values, **kw):
-        ids = []
-        size = len(values)
-        it = chunks(values, self._chunk_size, fmt=list)
+        if self._strategy == "multiprocessing" and not multi:
```
As discussed in DM: we should have two private implementations. The current one under `_create` and the new one in `_create_multiplrocessing` we call each depending on the options.

cawo-odoo commented:
added fixups to get rid of the scattered `_superchunk_size` logic and replaced it with local "batching" logic around `executor.map` calls. I'll fix py2 compat when #350 has finally been merged.
